### PR TITLE
Multi variable bit size

### LINKED ISSUE #78: Multiple variable sized items
A fairly common pattern is to have multiple length followed by variable length field within a packet. Come up with a probably inefficient, but working method of specifying this pattern.

### PR DESCRIPTION
**Pull request by ryanmelt**
closes #78

### CODE REVIEW

File: openc3/data/config/item_modifiers.yaml [hunk 1]
```diff
@@ -220,13 +220,3 @@ LIMITS_RESPONSE:
...
-KEY:
-  summary: Defines the key used to access this raw value in the packet.
-  description: Keys are often JsonPath or XPath strings
-  since: 5.0.10
-  parameters:
-    - name: Key string
-      required: true
-      description: The key to access this item
-      values: .+
-  example: KEY $.book.title
```
Comment (jmthomas) [inline openc3/data/config/item_modifiers.yaml:232]: Why was this removed? You still have an example in access_tlm.txt of KEY on an ITEM.

File: openc3/lib/openc3/accessors/binary_accessor.rb [hunk 2]
```diff
@@ -115,6 +115,149 @@ def self.raise_buffer_error(read_write, buffer, data_type, given_bit_offset, giv
...
+          case length_value
+          when 0
+            item.bit_size = 6
+          when 1
+            item.bit_size = 14
+          when 2
+            item.bit_size = 30
+          else
+            item.bit_size = 62
+          end
```
Comment (jmthomas) [inline openc3/lib/openc3/accessors/binary_accessor.rb:134]: Why are they all 2 bits short?

File: openc3/lib/openc3/accessors/binary_accessor.rb [hunk 3]
```diff
@@ -115,6 +115,149 @@ def self.raise_buffer_error(read_write, buffer, data_type, given_bit_offset, giv
...
+
+      # Recalculate bit offsets after this item
+      if adjustment != 0 and item.bit_offset >= 0
+        found = false
+        @packet.sorted_items.each do |sitem|
+          if sitem.data_type == :DERIVED or sitem.bit_offset < item.bit_offset
+            # Skip items before this item and derived items and items with negative bit offsets
+            next
+          end
+          if sitem != item
```
Comment (jmthomas) [inline openc3/lib/openc3/accessors/binary_accessor.rb:247]: The <=> operator in structure_item is based on bit_offset so you might be able to remove this check and just change the above check to `sitem.bit_offset <= item.bit_offset`

File: openc3/lib/openc3/packets/structure.rb [hunk 4]
```diff
@@ -337,6 +334,19 @@ def get_item(name)
...
       if @items[item.name]
         @items[item.name] = item
+        # Need to allocate space for the variable length item if its minimum size is greater than zero
+        if item.variable_bit_size
+          minimum_data_bits = 0
+          if (item.data_type == :INT or item.data_type == :UINT) and not item.original_array_size
+            # Minimum QUIC encoded integer
+            minimum_data_bits = 6
+          elsif item.variable_bit_size['length_value_bit_offset'] > 0
+            minimum_data_bits = item.variable_bit_size['length_value_bit_offset'] * item.variable_bit_size['length_bits_per_count']
```
Comment (jmthomas) [inline openc3/lib/openc3/packets/structure.rb:344]: This is effectively the STRING and BLOCK section?

File: openc3/spec/accessors/binary_accessor_spec.rb [hunk 5]
```diff
@@ -23,9 +23,137 @@
...
+        expect(item1.bit_offset).to eql 32
+        item2_length = @packet.append_item("item2_length", 16, :UINT)
+        expect(item2_length.bit_offset).to eql 32
+        expect(@packet.buffer).to eql "\x00\x00\x00\x04\x00\x00"
+        item2 = @packet.append_item("item2", 0, :STRING)
+        expect(@packet.buffer).to eql "\x00\x00\x00\x04\x00\x00"
+        @packet.write("item2_length", 8) # Lengths must be initialized to zero equivalent
+        item2.variable_bit_size = {'length_item_name' => 'item2_length', 'length_value_bit_offset' => -64, 'length_bits_per_count' => 8}
+        expect(item2.bit_offset).to eql 48
+        expect(@packet.buffer).to eql "\x00\x00\x00\x04\x00\x08"
```
Comment (jmthomas) [inline openc3/spec/accessors/binary_accessor_spec.rb:141]: I'm not sure what's going on with this test. How did we get the 4 and 8? Negative offset is a positive value? What happens if you have a value offset of +32?

**Review by jmthomas — COMMENTED** [review]
I still need to run locally and see about the coverage and additional testing

File: openc3/data/config/item_modifiers.yaml (hunk 1 above)
Comment (ryanmelt) [reply openc3/data/config/item_modifiers.yaml:232]: It was just remove from item_modifiers because it was a dup (I think).  It is also in param_item_modifiers.

File: openc3/lib/openc3/accessors/binary_accessor.rb (hunk 2 above)
Comment (ryanmelt) [reply openc3/lib/openc3/accessors/binary_accessor.rb:134]: Its the QUIC variable size integer format.  See: https://datatracker.ietf.org/doc/html/rfc9000#name-variable-length-integer-enc

File: openc3/spec/accessors/binary_accessor_spec.rb (hunk 5 above)
Comment (ryanmelt) [reply openc3/spec/accessors/binary_accessor_spec.rb:141]: This test is handling the case where the value field includes itself in the count (so that has to be subtracted out)
A positive offset would be something like CCSDS where a value of 0 = 1 byte = 8 bits.  Don't have a test for that yet.

File: openc3/data/config/item_modifiers.yaml (hunk 1 above)
Comment (jmthomas) [reply openc3/data/config/item_modifiers.yaml:232]: Right and that's used by both ... not sure why it didn't get duplicated in the output though

File: openc3/lib/openc3/accessors/binary_accessor.rb (hunk 2 above)
Comment (jmthomas) [reply openc3/lib/openc3/accessors/binary_accessor.rb:134]: That's above and beyond ... I thought you were just going to do strings, lol

File: openc3/lib/openc3/packets/packet_config.rb [hunk 6]
```diff
@@ -469,6 +473,13 @@ def process_current_packet(parser, keyword, params)
         @current_packet.hidden = true
         @current_packet.disabled = true
 
+      when 'VIRTUAL'
+        usage = "#{keyword}"
+        parser.verify_num_parameters(0, 0, usage)
+        @current_packet.hidden = true
+        @current_packet.disabled = true
+        @current_packet.virtual = true
```
Comment (jmthomas) [inline openc3/lib/openc3/packets/packet_config.rb:481]: What is a VIRTUAL packet and how is it going to be used?

File: openc3/lib/openc3/conversions/object_read_conversion.rb [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# This file may also be used under the terms of a commercial license
+# if purchased from OpenC3, Inc.
+
+require 'openc3/conversions/conversion'
+
+module OpenC3
+  class ObjectReadConversion < Conversion
```
Comment (jmthomas) [inline openc3/lib/openc3/conversions/object_read_conversion.rb:22]: What is this for?

File: openc3/lib/openc3/conversions/object_write_conversion.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# This file may also be used under the terms of a commercial license
+# if purchased from OpenC3, Inc.
+
+require 'openc3/conversions/object_read_conversion'
+
+module OpenC3
+  class ObjectWriteConversion < Conversion
```
Comment (jmthomas) [inline openc3/lib/openc3/conversions/object_write_conversion.rb:22]: What is this for?

File: openc3/lib/openc3/packets/packet_config.rb (hunk 6 above)
Comment (ryanmelt) [reply openc3/lib/openc3/packets/packet_config.rb:481]: VIRTUAL packs aren't involved in identification.  So it provides a way to declare embedded structures that can be used within packets.   Effectively packets in packets.

File: openc3/lib/openc3/conversions/object_read_conversion.rb (hunk 7 above)
Comment (ryanmelt) [reply openc3/lib/openc3/conversions/object_read_conversion.rb:22]: Used with virtual.  This is a new conversion that references a packet structure, and then dumps the BLOCK data into that packet, and returns a hash of key/value pairs generated using the packet structure.

File: openc3/lib/openc3/conversions/object_write_conversion.rb (hunk 8 above)
Comment (ryanmelt) [reply openc3/lib/openc3/conversions/object_write_conversion.rb:22]: Used with virtual, to write a hash of items, into a packet, and then fill in a BLOCK item.

File: openc3/lib/openc3/accessors/binary_accessor.rb [hunk 9]
```diff
@@ -108,13 +106,171 @@ def self.raise_buffer_error(read_write, buffer, data_type, given_bit_offset, giv
...
+          # We added bytes to the end so now we have to shift the buffer over
+          #   NOTE: buffer[offset, length]
+          # We copy to the shifted offset location with the remaining buffer length
+          buffer[item_offset + bytes, buffer.length - (item_offset + bytes)] =
+              # We copy from the original offset location with the original length minus the offset
+              buffer[item_offset, original_length - item_offset]
+        elsif bytes < 0
+          # Remove extra bytes because we're adjusting smaller
+          buffer[item_offset + -bytes, -bytes] = ''
+        end
```
Comment (jmthomas) [inline openc3/lib/openc3/accessors/binary_accessor.rb:226]: This is my new buffer adjustment logic

File: openc3/lib/openc3/accessors/binary_accessor.rb [hunk 10]
```diff
@@ -108,13 +106,171 @@ def self.raise_buffer_error(read_write, buffer, data_type, given_bit_offset, giv
...
+          buffer << ("\000" * bytes)
+          # We added bytes to the end so now we have to shift the buffer over
+          #   NOTE: buffer[offset, length]
+          # We copy to the shifted offset location with the remaining buffer length
+          buffer[item_offset + bytes, buffer.length - (item_offset + bytes)] =
+              # We copy from the original offset location with the original length minus the offset
+              buffer[item_offset, original_length - item_offset]
+        elsif bytes < 0
+          # Remove extra bytes because we're adjusting smaller
+          buffer[item_offset + -bytes, -bytes] = ''
```
Comment (ryanmelt) [inline openc3/lib/openc3/accessors/binary_accessor.rb:225]: This doesn't seem right.  Let's say we got 7 bytes smaller.   
buffers[0 + 7, 7] = "" will destroy a bunch of real data.
Should it always be: buffer[item_offset + 1, -bytes] = '' ?

**Review by jmthomas — APPROVED** [review]

**Review by ryanmelt — COMMENTED** [review]
Would like tests for the new conversion classes, and I think packet_config_spec should also include some new tests.

**Comment by jmthomas** [comment]
> Would like tests for the new conversion classes, and I think packet_config_spec should also include some new tests.

Tests added and packet_config.rb is at 96.58%

**Comment by ryanmelt** [comment]
Looks good!
